### ISSUE TITLE
no-unused-vars false positive

### ISSUE THREAD

**Issue by Robloche** [eslint/eslint#12171]
* **ESLint Version: 6.2.2**
* **Node Version: 8.12.0**
* **npm Version: 6.11.2**

**What parser (default, Babel-ESLint, etc.) are you using?**
Babel-ESLint.

<details>
<summary>Configuration</summary>

```js
{
  "parser": "babel-eslint",
  "plugins": [
    "react",
    "flowtype"
  ],
  "settings": {
    "react": {
      "version": "detect"
    }
  },
  "env": {
    "browser": true,
    "node": true,
    "es6": true
  },
  "rules": {
    "flowtype/define-flow-type": 1,
    "flowtype/use-flow-type": 1,
    "max-len": [1, 200, 2, {"ignoreComments": true}],
    "max-lines": ["error", {"skipBlankLines": true, "skipComments": true, "max": 600}],
    "max-lines-per-function": ["error", {"skipBlankLines": true, "skipComments": true, "max": 100}],
    "quotes": ["error", "single", { "avoidEscape": true }],
    "jsx-quotes": ["error", "prefer-single"],
    "react/jsx-filename-extension": [1, { "extensions": [".js", ".jsx"] }],
    "indent": ["error", 2, { "SwitchCase": 1 }],
    "space-before-function-paren": ["error", "never"],
    "class-methods-use-this": "off",
    "padded-blocks": ["error", { "blocks": "never", "classes": "always", "switches": "never" }],
    "react/jsx-indent": ["error", 2],
    "react/jsx-indent-props": ["error", 2],
    "react/jsx-closing-bracket-location": [1, {"selfClosing": "after-props", "nonEmpty": "after-props"}],
    "react/jsx-closing-tag-location": "off",
    "react/jsx-wrap-multilines": "off",
    "react/jsx-no-literals": "off",
    "react/jsx-max-depth": [2, { "max": 10 }],
    "react/jsx-one-expression-per-line": [0],
    "dot-location": [2, "property"],
    "curly": "off",
    "quote-props": ["error", "as-needed"],
    "no-invalid-this": "off",
    "react/no-set-state": "off",
    "no-ternary": "off",
    "multiline-ternary": "off",
    "arrow-body-style": ["error", "as-needed", { "requireReturnForObjectLiteral": true }],
    "one-var": "off",
    "block-spacing": ["error", "never"],
    "brace-style": ["error", "1tbs", { "allowSingleLine": true }],
    "react/jsx-no-bind": ["error", { "ignoreRefs": true }],
    "react/sort-comp": ["error", { "order": ["type-annotations", "static-methods", "lifecycle", "everything-else", "render"] }],
    "react/forbid-component-props": "off",
    "max-params": "off",
    "no-negated-condition": "off",
    "lines-around-comment": ["error", { "beforeBlockComment": false }],
    "no-plusplus": ["error", { "allowForLoopAfterthoughts": true }],
    "max-statements": "off",
    "no-magic-numbers": ["error", { "ignore": [0, 1, -1] }],
    "function-paren-newline": ["error", "consistent"],
    "newline-per-chained-call": "off",
    "id-length": "off",
    "no-process-env": "off",
    "no-extra-parens": ["error", "functions"],
    "no-use-before-define": ["error", { "functions": true, "classes": true, "variables": false }],
    "no-undefined": "off",
    "func-style": ["error", "declaration", { "allowArrowFunctions": true }],
    "function-call-argument-newline": "off",
    "no-loop-func": "error",
    "callback-return": "off",
    "no-undef-init": "off",
    "react/prefer-stateless-function": "off",
    "implicit-arrow-linebreak": "off",
    "react/no-danger": "off",
    "complexity": ["error", 25],
    "max-depth": ["error", 6],
    "no-warning-comments": 0,
    "prefer-named-capture-group": "off"
  },
  "extends": ["eslint:all", "plugin:react/all"]
}
```

</details>

I have the following chunks of code which were not raising any error before I update to eslint 6.2.2:

```js
for (const arg of args) {
    const [key, value] = arg.split('=');
    // Code using 'key' and 'value'
  }
```

```js
for (const key of keys) {
    const {[key]: val} = map;
    // Code using 'val'
}
```

```bash
eslint . --ext .js,.jsx --cache
```

**What did you expect to happen?**
No warning or error, just like before the update.

**What actually happened? Please include the actual, raw output from ESLint.**
```
[eslint   ] C:\path\file1.js
[eslint   ]   42:14  error  'arg' is defined but never used  no-unused-vars
[eslint   ]
[eslint   ] C:\path\file2.js
[eslint   ]   74:14  error  'key' is defined but never used  no-unused-vars
[eslint   ]
[eslint   ] ✖ 2 problems (2 errors, 0 warnings)
```

**Comment by Robloche** [comment]
OK, thanks!
